### PR TITLE
Support Configurable Composite Alpha Modes, Transparency, & Renderer Options

### PR DESCRIPTION
**Pull request by varappdev**
## Summary

This pull request introduces the ability to configure composite alpha modes and base colors across various renderers. Previously, the composite alpha mode was hardcoded (e.g. to `Auto`, which often defaults to `Opaque`), making alpha-blending and transparency difficult or impossible.

We now expose configurable option structs and allow specifying the background/clear colors and composite alpha modes for Vello, Vello Hybrid, Skia, and Pixels (CPU) renderers. Additionally, we make key options structs future-proof by marking them as `#[non_exhaustive]`.

> [!NOTE]
> `softbuffer` does not support transparency natively (yet), so these alpha-compositing features will apply only to the GPU and Pixels-based paths.
>
> A small adjustment will need to be made in `dioxus-native` to use struct updates or `Default::default()` when constructing renderer options to accommodate the `#[non_exhaustive]` attribute.

### 1. Transparency & Alpha Compositing Control

- **Vello Renderers**: Added `composite_alpha_mode` option to vello's `window_renderer.rs` and vello-hybrid's `window_renderer.rs`. This controls the surface compositing behavior of the WGPU surface inside the respective window renderers.
- **Pixels CPU / Raster Renderers**: Added configuration options via the `PixelsRendererOptions` struct inside `lib.rs` to customize the `base_color` and `composite_alpha_mode` using the `PixelsBuilder` API.
- **Skia Renderer**: Added configuration options via the `SkiaRendererOptions` inside `window_renderer.rs` to customize the `base_color` only as Skia handles transparency by default.
- **Skia Raster Renderer**: Clears the canvas with `Color::TRANSPARENT` in `image_renderer.rs` by default to preserve alpha channels in rendered outputs.

### 2. Extensibility & Future Proofing

- Marked options in vello's `window_renderer.rs` and vello-hybrid's `window_renderer.rs` as `#[non_exhaustive]` to allow adding more configuration fields in the future without breaking user builds.

### 3. Crate-specific Options & Structs

- **`anyrender_vello_cpu`**: Exported `VelloCpuRendererOptions` as an alias for `PixelsRendererOptions`.
- **`anyrender_skia`**: Exported `SkiaRasterRendererOptions` as an alias for `PixelsRendererOptions`.

## Breaking Changes

- The configuration structs `VelloRendererOptions` and `VelloHybridRendererOptions` are now `#[non_exhaustive]`. Constructing these structs directly without using `Default::default()` or struct update syntax will cause compilation errors.

### CODE REVIEW

**Review by nicoburns — COMMENTED** [review]
The implementation of this seems fine, but it's a bit unfortunate that each renderer has a different API. Do you think it might be possible to unify the APIs somehow?

I'm thinking about how we'd expose that to users at the higher level, and it seems like they'll need be some conversion between base color and opacity there anyway...

**Comment by varappdev** [comment]
> The implementation of this seems fine, but it's a bit unfortunate that each renderer has a different API. Do you think it might be possible to unify the APIs somehow?

Nothing elegant comes to mind right away, but let me take another look at it later today to see if I can sketch out a better approach.

> I'm thinking about how we'd expose that to users at the higher level, and it seems like they'll need be some conversion between base color and opacity there anyway...

Can you clarify what you mean by the conversion? Just want to make sure we're on the same page, since both Peniko and Skia colour types already include alpha.

**Comment by varappdev** [comment]
How do we feel about this? Base colour and composite alpha mode seem to be the most shared fields so I thought they are the ones that make the most sense to share.

As for providing a base colour to vello hybrid I looked into it and it is hardcoded to fill with transparent. I could somewhat implement this option such that if a colour but transparent is defined it fills the entire surface with the provided colour.

**Comment by nicoburns** [comment]
I think the constraints I am trying to fulfil here are:

- The user can pass the same unified config and get the same output regardless of renderer (excl. softbuffer where alpha isn't supported yet).
- Ideally everything is setup for optimal performance (or close to) on all renderers (but this is less important than identical output)

**Comment by varappdev** [comment]
I've implemented composite alpha mode for Skia and base colour support for `vello_hybrid` and `softbuffer`.

Skia just required some parameter passing to the backends. However, `vello_hybrid` needed to be manually told to do an initial fill with a colour, and `softbuffer` required some manual pixel math to prevent rendering artifacts with partial alpha values. 

Just a heads-up: I don't have access to a Mac, so I wasn't able to test the behavior on the Metal Skia backend.

Edit: nevermind skia on opengl refuses to not be transparent I'll have to look into that further.

**Comment by varappdev** [comment]
OpenGL should be behaving properly with transparency now. 

I think we've covered all the shared options between the renderers at this point. Let me know if there's anything else you need before we merge!

**Comment by nicoburns** [comment]
This is now looking pretty good to me from a code review POV. But I want to wire it through Blitz / Dioxus Native before merging to make sure that it actually does what we want it to do end-to-end (and allow me to actually test it).

**Comment by nicoburns** [comment]
@varappdev I've tried adding this to Blitz (https://github.com/DioxusLabs/blitz/pull/517). And it basically works, however there is an issue with selecting alpha mode: my machine (macOS) only has [`Opaque`, `PostMultiplied`] and not `PreMultiplied`, so it crashes if I specify `PreMultiplied`, but presumably other machines would have `PreMultiplied` but not `PostMultiplied` so they would crash the other way around.

So I think we need to either:
- Make the setting a closure (probably `Box<dyn FnMut(&[AlphaMode]>`) so that the user can choose from the list of available modes
- OR have some kind of more generic just `Alpha | Opaque` and have the system choose.

**Comment by varappdev** [comment]
> @varappdev I've tried adding this to Blitz ([DioxusLabs/blitz#517](https://github.com/DioxusLabs/blitz/pull/517)). And it basically works, however there is an issue with selecting alpha mode: my machine (macOS) only has [`Opaque`, `PostMultiplied`] and not `PreMultiplied`, so it crashes if I specify `PreMultiplied`, but presumably other machines would have `PreMultiplied` but not `PostMultiplied` so they would crash the other way around.
> 
> So I think we need to either:
> 
>    * Make the setting a closure (probably `Box<dyn FnMut(&[AlphaMode]>`) so that the user can choose from the list of available modes
> 
>    * OR have some kind of more generic just `Alpha | Opaque` and have the system choose.

I was actually aware of that discrepancy, as on the Linux Wayland side we only have `Opaque` and `PreMultiplied`. I originally thought leaving it up to the user to choose the corresponding mode for their platform would work, but I see how that causes immediate friction. 

I'll go ahead and implement a mix of both your suggestions: I'll add a generic `Alpha | Opaque` selection to the shared options struct, and use the closure approach for the individual backend options struct so users can still fine-tune it if needed.

**Comment by nicoburns** [comment]
@varappdev I've looked into this some more and:

- Vello Classic (`vello`) outputs in straight (unpremultiplied) alpha
- Vello Hybrid (`vello_hybrid`) outputs in premultiplied alpha
- (Vello CPU (`vello_cpu`) also outputs premultiplied alpha)

So I think we could probably hardcode a preferred alpha mode in each of those renders (if user requests alpha support)? 

(what do you think? do we still need the closure-based override?)

I also have some code for converting between the different alpha formats in the case that available alpha mode doesn't match the output mode of the render in use:

- https://github.com/DioxusLabs/anyrender/tree/alpha-conversions

It makes use of support for rendering to an intermediate texture that we already had in place for Vello Classic (because it can't render directly to the surface on all platforms anyway (because it needs to be able to read from the surface which isn't always supported)).

**Comment by varappdev** [comment]
> @varappdev I've looked into this some more and:
> 
>     * Vello Classic (`vello`) outputs in straight (unpremultiplied) alpha
> 
>     * Vello Hybrid (`vello_hybrid`) outputs in premultiplied alpha
> 
>     * (Vello CPU (`vello_cpu`) also outputs premultiplied alpha)
> 
> 
> So I think we could probably hardcode a preferred alpha mode in each of those renders (if user requests alpha support)?
> 
> (what do you think? do we still need the closure-based override?)
> 
> I also have some code for converting between the different alpha formats in the case that available alpha mode doesn't match the output mode of the render in use:
> 
>     * https://github.com/DioxusLabs/anyrender/tree/alpha-conversions
> 
> 
> It makes use of support for rendering to an intermediate texture that we already had in place for Vello Classic (because it can't render directly to the surface on all platforms anyway (because it needs to be able to read from the surface which isn't always supported)).

Upon further investigation on my end, it seems like virtually all platforms actually expect `PreMultiplied` alpha—including macOS. It turns out that when the `wgpu` developers were writing the Metal backend, they accidentally misreported that macOS only handles `PostMultiplied`, and that quirk has just stuck around.

Because of this, I completely agree that we don't need the closure-based override. To be honest, that's a bit of a relief on my end, as `pixels` was acting up a bit on the `wgpu` Gles backend when I was looking into hooking up the closures!

Hardcoding the preferred alpha modes based on the renderer makes perfect sense. We can safely target:
* **Metal (all apple platforms but watchOS):** `PostMultiplied` (straight alpha) when the user asks for alpha.
* **All other platforms:** `PreMultiplied`.

This is great because it means we only ever have to handle alpha conversion one way (converting `PostMultiplied`/straight alpha to `PreMultiplied`). Your intermediate texture approach in that branch looks like the perfect way to handle this!

**Comment by nicoburns** [comment]
> Upon further investigation on my end, it seems like virtually all platforms actually expect PreMultiplied alpha—including macOS. It turns out that when the wgpu developers were writing the Metal backend, they accidentally misreported that macOS only handles PostMultiplied, and that quirk has just stuck around.

Do you have a reference for this. It seems very surprising that this wouldn't get fixed if it's supported by the underlying platform.

> This is great because it means we only ever have to handle alpha conversion one way (converting PostMultiplied/straight alpha to PreMultiplied).

Do we not still need PreMultiplied -> PostMultiplied conversion for vello_hybrid on macOS in this setup?

We can safely target:

- Metal (all apple platforms but watchOS): PostMultiplied (straight alpha) when the user asks for alpha.
- All other platforms: PreMultiplied.

I think this is probably fine, as it seems like the available modes are consistent by platform, but maybe we should have a preferred mode based on the mode the  *renderer* (Vello, Vello Hybrid) natively outputs, and then fallback to the other mode from the list of available modes if the preferred mode is not available.

Then have the conversion just go off what mode what actually chosen (and what mode the renderer outputs).

**Comment by varappdev** [comment]
> Do you have a reference for this. It seems very surprising that this wouldn't get fixed if it's supported by the underlying platform.

If you look at this [line in `wgpu-hal`](https://github.com/gfx-rs/wgpu/blob/a9beb0dc1f60338df07a4d43d9ec0e152d03611b/wgpu-hal/src/metal/surface.rs#L271), they hardcode `PostMultiplied` to control the `isOpaque` property of the `CALayer` (`CAMetalLayer` inherits from it). The [Apple Developer Docs for `isOpaque`](https://developer.apple.com/documentation/quartzcore/calayer/isopaque) outline the property itself, but they don't actually explain the underlying composition mechanics. However, as this [classic Stack Overflow answer](https://stackoverflow.com/a/51414692) points out from Apple engineer insights, macOS fundamentally expects **premultiplied** alpha for transparent layers. 

So the fact that `wgpu` maps this to `PostMultiplied` is an internal quirk from when the Metal backend was written. My point with these sources is that Apple does, in fact, expect premultiplied alpha.

> Do we not still need PreMultiplied -> PostMultiplied conversion for vello_hybrid on macOS in this setup?

Probably not if that QuartzCore logic holds true, though since I don't have a Mac to hand, I unfortunately can't verify the end-to-end behaviour myself.

> I think this is probably fine, as it seems like the available modes are consistent by platform, but maybe we should have a preferred mode based on the mode the *renderer* (Vello, Vello Hybrid) natively outputs, and then fallback to the other mode from the list of available modes if the preferred mode is not available.
> 
> Then have the conversion just go off what mode what actually chosen (and what mode the renderer outputs).

I understand wanting a fallback mechanism to try and bypass a premultiply pass where a renderer natively outputs straight alpha. The issue is that `PostMultiplied` support is incredibly spotty and unreliable across platforms. 

The only platform that properly and reliably implements `PostMultiplied` is Windows DXGI (specifically if `flip_model` is set to `FLIP_SEQUENTIAL` or `FLIP_DISCARD`). For Vulkan and OpenGL on Windows, it completely depends on whether the graphics drivers implement it. On Linux with X11, you're entirely at the mercy of the drivers as well, where it's very often mishandled. Meanwhile, Linux Wayland flat out refuses to hand you a `PostMultiplied` option in almost every instance, and Apple (allegedly) does not handle it either. 

Because `PostMultiplied` support is so broken or non-existent outside of very specific environments, trying to dynamically fall back might introduce a lot of fragile edge cases compared to just hardcoding the known stable paths per platform.

If you want to verify what macOS actually expects end-to-end, a quick way to test it is to check a CSS gradient fading into transparency against a transparent window background. 

I actually ran into this exact issue by accident on Linux Wayland (which strictly expects `PreMultiplied`). Because Vello Classic outputs straight alpha by default, the CSS gradients against transparency ended up looking way too bright and blown out because the unmultiplied colors were bleeding into the alpha channel blend. 



You can see the exact RSX and CSS I used to catch this here: https://git.varapp.dev/varappdev/hshell/src/commit/e88c03ff6c4edbf03cec90b750acdf8009db1aff/hshell/src/main.rs#L29.

If macOS behaves the same way when forced into `PreMultiplied` or `PostMultiplied`, the visual artifacts will be glaring:
* **Straight alpha treated as premultiplied:** The fading edges will have a bright, distinct "halo" or fringe (exactly like the blown-out gradients on Wayland).
* **Premultiplied treated as straight:** The edges will drop off into a dark, muddy, or black border right before hitting 0% opacity.

**Comment by nicoburns** [comment]
I've done some investigating into this (including running tests on my local machine), and I'm pretty convinced that you're correct about PreMultiplied vs PostMultiplied on macOS. I've opened https://github.com/gfx-rs/wgpu/issues/9896 to raise it with WGPU developers.

Regarding choosing the alpha mode (when user requests alpha), could we have a very simple algorithm:

- Choose `PreMultiplied` if it's available
- Else fallback to `PostMultiplied` (+ hack on apple platforms to treat `PostMultipled` as `PreMultiplied`)

**Comment by varappdev** [comment]
> I've done some investigating into this (including running tests on my local machine), and I'm pretty convinced that you're correct about PreMultiplied vs PostMultiplied on macOS. I've opened [gfx-rs/wgpu#9896](https://github.com/gfx-rs/wgpu/issues/9896) to raise it with WGPU developers.
> 
> Regarding choosing the alpha mode (when user requests alpha), could we have a very simple algorithm:
> 
>     * Choose `PreMultiplied` if it's available
> 
>     * Else fallback to `PostMultiplied` (+ hack on apple platforms to treat `PostMultipled` as `PreMultiplied`)

Glad to hear the testing confirmed it! Thanks for opening that upstream issue as well.

I've finished implementing the alpha mode selection logic using that algorithm, along with adding some `// TODO` comments to serve as a reminder for when the `wgpu` bug eventually gets resolved. 

To make this robust, I also added a method to the window renderer trait that reports exactly which alpha mode was ultimately chosen. This way, on the off-chance a platform actually lands on `PostMultiplied`, renderers that output straight alpha natively can skip their conversion pass entirely, while the other renderers know they need to start converting their output.

**Comment by varappdev** [comment]
And now I have pulled your earlier commit from the alpha-conversion and integrated it into the alpha mode selection.

File: crates/anyrender_vello/src/window_renderer.rs [hunk 1]
```diff
@@ -215,6 +297,54 @@ impl WindowRenderer for VelloWindowRenderer {
...
+                    TextureConfiguration {
+                        usage: TextureUsages::STORAGE_BINDING | TextureUsages::TEXTURE_BINDING,
+                        format: TextureFormat::Rgba8Unorm,
+                        alpha_conversion: Some(AlphaConversion::Premultiply),
+                    }
+                });
+            // TODO: Remove below once gfx-rs/wgpu#9896 gets fixed
+            #[cfg(target_vendor = "apple")]
+            let texture_config = Some(TextureConfiguration {
+                usage: TextureUsages::STORAGE_BINDING,
```
Comment (nicoburns) [inline crates/anyrender_vello/src/window_renderer.rs]: This needs to be `TextureUsages::STORAGE_BINDING | TextureUsages::TEXTURE_BINDING`

File: crates/anyrender_vello/src/window_renderer.rs [hunk 2]
```diff
@@ -215,6 +297,54 @@ impl WindowRenderer for VelloWindowRenderer {
...
+                );
+                composite_alpha_mode = alpha_modes
+                    .first()
+                    .copied()
+                    .expect("Surface didn't report any alpha modes");
+            }
+
+            #[cfg(not(target_vendor = "apple"))]
+            let texture_config =
+                (composite_alpha_mode == CompositeAlphaMode::PreMultiplied).then(|| {
```
Comment (nicoburns) [inline crates/anyrender_vello/src/window_renderer.rs]: This is wrong. The intermediate texture is always required for vello classic.

File: crates/anyrender/src/types.rs [hunk 3]
```diff
@@ -123,3 +123,89 @@ impl<'a> From<BrushRef<'a>> for PaintRef<'a> {
...
+
+    fn try_from(value: u8) -> Result<Self, Self::Error> {
+        match value {
+            0 => Ok(Self::PreMultiplied),
+            1 => Ok(Self::PostMultiplied),
+            2 => Ok(Self::Opaque),
+            _ => Err(()),
+        }
+    }
+}
```
Comment (nicoburns) [inline crates/anyrender/src/types.rs]: What is this for?

File: crates/anyrender_skia/src/window_renderer.rs [hunk 4]
```diff
@@ -85,6 +163,43 @@ impl WindowRenderer for SkiaWindowRenderer {
...
+        {
+            any_backend
+                .downcast_ref::<crate::vulkan::VulkanBackend>()
+                .map(|b| b.current_alpha_mode())
+        }
+        #[cfg(not(feature = "vulkan"))]
+        {
+            None
+        }
+    }
```
Comment (nicoburns) [inline crates/anyrender_skia/src/window_renderer.rs]: Rather than all this downcasting, could we add a `current_alpha_mode` method to the `SkiaBackend` trait, and implement it for all Skia backends.

File: crates/anyrender/src/types.rs [hunk 5]
```diff
@@ -123,3 +123,89 @@ impl<'a> From<BrushRef<'a>> for PaintRef<'a> {
...
+
+/// Error returned when converting a unified configuration to a backend-specific one.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum ConfigError {
+    /// The target renderer does not support this configuration option.
+    UnsupportedField(&'static str),
+}
```
Comment (nicoburns) [inline crates/anyrender/src/types.rs:142]: This appears to be unused. Leftover code?

File: crates/softbuffer_window_renderer/src/lib.rs [hunk 6]
```diff
@@ -128,13 +203,48 @@ impl<Renderer: ImageRenderer> WindowRenderer for SoftbufferWindowRenderer<Render
...
+
+                    out_r = (r * a * 255 + base_r * blend_factor) / denom;
+                    out_g = (g * a * 255 + base_g * blend_factor) / denom;
+                    out_b = (b * a * 255 + base_b * blend_factor) / denom;
+                } else {
+                    // Both source and base are completely transparent (out_a == 0)
+                    out_r = 0;
+                    out_g = 0;
+                    out_b = 0;
+                }
```
Comment (nicoburns) [inline crates/softbuffer_window_renderer/src/lib.rs]: I think this is going to be too expensive (and softbuffer doesn't support transparency anyway...). Could we wrap the `draw_fn` in a function that pushes a rectangular fill and then calls the `draw_fn` instead? Or else just abandon matching the other renderers with softbuffer for now.

**Review by nicoburns — CHANGES_REQUESTED** [review]
Sorry for the delay on this. But I have requested some changes.

File: crates/anyrender_vello/src/window_renderer.rs (hunk 1 above)
Comment (varappdev) [reply crates/anyrender_vello/src/window_renderer.rs]: Fixed

File: crates/anyrender_vello/src/window_renderer.rs (hunk 2 above)
Comment (varappdev) [reply crates/anyrender_vello/src/window_renderer.rs]: Fixed as well

File: crates/anyrender/src/types.rs (hunk 3 above)
Comment (varappdev) [reply crates/anyrender/src/types.rs]: Not needed anymore, realized we don't actually need an additional trait function or this struct. It was from when I was experimenting with using Atomic U8s to store the alpha mode.

File: crates/anyrender_skia/src/window_renderer.rs (hunk 4 above)
Comment (varappdev) [reply crates/anyrender_skia/src/window_renderer.rs]: Not needed anymore, realized we don't actually need an additional trait function.

File: crates/anyrender/src/types.rs (hunk 5 above)
Comment (varappdev) [reply crates/anyrender/src/types.rs:142]: Now used when someone asks for transparency in softbuffer window renderer.

File: crates/softbuffer_window_renderer/src/lib.rs (hunk 6 above)
Comment (varappdev) [reply crates/softbuffer_window_renderer/src/lib.rs]: Done, we now wrap draw_fn.

**Comment by varappdev** [comment]
I have also added a unpremultiply step to the skia vulkan backend if it encounters PostMultiplied for when it is eventually working.

**Comment by varappdev** [comment]
@nicoburns I think you might want to check why the build job is stuck before all the runner minutes are burned.

File: crates/anyrender/src/null_backend.rs [hunk 7]
```diff
@@ -46,6 +46,10 @@ impl WindowRenderer for NullWindowRenderer {
 
     fn set_size(&mut self, _width: u32, _height: u32) {}
 
+    // fn current_alpha_mode(&self) -> Option<CurrentCompositeAlphaMode> {
+    //     Some(CurrentCompositeAlphaMode::Opaque)
+    // }
+
```
Comment (nicoburns) [inline crates/anyrender/src/null_backend.rs]: Please remove this commented code

File: crates/anyrender_vello/src/window_renderer.rs [hunk 8]
```diff
@@ -287,6 +412,37 @@ impl WindowRenderer for VelloWindowRenderer {
...
+    //             CompositeAlphaMode::PostMultiplied => {
+    //                 Some(CurrentCompositeAlphaMode::PreMultiplied)
+    //             }
+    //             CompositeAlphaMode::Auto
+    //             | CompositeAlphaMode::Opaque
+    //             | CompositeAlphaMode::Inherit => Some(CurrentCompositeAlphaMode::Opaque),
+    //         }
+    //     }
+    // }
+
```
Comment (nicoburns) [inline crates/anyrender_vello/src/window_renderer.rs]: Please remove commented code

File: crates/anyrender_vello/src/window_renderer.rs [hunk 9]
```diff
@@ -215,6 +296,52 @@ impl WindowRenderer for VelloWindowRenderer {
...
+                format: TextureFormat::Rgba8Unorm,
+                alpha_conversion: (composite_alpha_mode == CompositeAlphaMode::PreMultiplied)
+                    .then_some(AlphaConversion::Premultiply),
+            });
+            // TODO: Remove below once gfx-rs/wgpu#9896 gets fixed
+            #[cfg(target_vendor = "apple")]
+            let texture_config = Some(TextureConfiguration {
+                usage: TextureUsages::STORAGE_BINDING | TextureUsages::TEXTURE_BINDING,
+                format: TextureFormat::Rgba8Unorm,
+                alpha_conversion: Some(AlphaConversion::Premultiply),
```
Comment (nicoburns) [inline crates/anyrender_vello/src/window_renderer.rs]: For apple, `alpha_conversion` should only be PreMultiply if transparency is enabled. If we have an Opaque texture/surface then this should be a simple copy/blit.

File: crates/anyrender_vello/src/window_renderer.rs [hunk 10]
```diff
@@ -215,6 +296,52 @@ impl WindowRenderer for VelloWindowRenderer {
                 .expect("Error creating DeviceHandle"),
             };
 
+            let adapter = &device_handle.adapter;
+            let caps = surface.get_capabilities(adapter);
+            let mut alpha_modes = caps.alpha_modes;
+            let found = alpha_modes.iter().find(|m| **m == composite_alpha_mode);
```
Comment (nicoburns) [inline crates/anyrender_vello/src/window_renderer.rs]: Non-blocking nit: You could use `.contains()` here rather than `.find()`. Same in vello_hybrid.

**Review by nicoburns — CHANGES_REQUESTED** [review]
I think we're getting close. A few small changes requested.

File: crates/softbuffer_window_renderer/src/lib.rs [hunk 11]
```diff
@@ -40,6 +92,42 @@ impl<Renderer: ImageRenderer> SoftbufferWindowRenderer<Renderer> {
...
+        config: impl TryInto<SoftbufferRendererOptions, Error = impl std::error::Error>,
+    ) -> Self {
+        Self {
+            render_state: RenderState::Suspended,
+            window_handle: None,
+            renderer: Renderer::new(0, 0),
+            buffer: Vec::new(),
+            config: config
+                .try_into()
+                .expect("Invalid Softbuffer renderer configuration"),
```
Comment (nicoburns) [inline crates/softbuffer_window_renderer/src/lib.rs]: The `expect` here makes using TryInto a bit pointless...

File: crates/anyrender/src/null_backend.rs (hunk 7 above)
Comment (varappdev) [reply crates/anyrender/src/null_backend.rs]: Sorry about that i missed it.

File: crates/anyrender_vello/src/window_renderer.rs (hunk 8 above)
Comment (varappdev) [reply crates/anyrender_vello/src/window_renderer.rs]: Sorry about that i missed it.

File: crates/anyrender_vello/src/window_renderer.rs (hunk 9 above)
Comment (varappdev) [reply crates/anyrender_vello/src/window_renderer.rs]: I have found out that if the alpha mode is set to Opaque and the clear colour is set to transparent it leads to some issues. for example, if the texture config is the following
```rs
#[cfg(not(target_vendor = "apple"))]
let texture_config = Some(TextureConfiguration {
    usage: TextureUsages::STORAGE_BINDING | TextureUsages::TEXTURE_BINDING,
    format: TextureFormat::Rgba8Unorm,
    alpha_conversion: None,
});
```
and I configure the renderer like so
```
let mut options = VelloRendererOptions::default()
    .base_color(peniko::Color::TRANSPARENT)
    .composite_alpha_mode(anyrender::CompositeAlphaMode::Opaque);
let mut renderer = VelloWindowRenderer::with_options(options);
```
it leads to strange issues like this (at least on linux) (the sides should be a gradient not colour)
image

Then again if someone asks for a transparent clear colour on opaque alpha mode maybe we shouldn't account for that.

File: crates/anyrender_vello/src/window_renderer.rs (hunk 10 above)
Comment (varappdev) [reply crates/anyrender_vello/src/window_renderer.rs]: I forgot that function exists for a moment my mistake.

File: crates/softbuffer_window_renderer/src/lib.rs (hunk 11 above)
Comment (varappdev) [reply crates/softbuffer_window_renderer/src/lib.rs]: made it return a result instead

Comment (varappdev) [reply crates/softbuffer_window_renderer/src/lib.rs]: and also made the rest into a simple Into

File: crates/anyrender_vello/src/window_renderer.rs (hunk 9 above)
Comment (nicoburns) [reply crates/anyrender_vello/src/window_renderer.rs]: > Then again if someone asks for a transparent clear colour on opaque alpha mode maybe we shouldn't account for that.

Yeah, I think we can probably leave this as a "just don't do this"

**Comment by varappdev** [comment]
This should be it I think. Let me know if you want the commits squashed in any way. I avoided force pushes to not break your changes since view.

**Comment by varappdev** [comment]
Thanks so much for getting those last commits over the line and merging this. 

Since this was my first proper open-source contribution, I really appreciate your patience through the review process, especially with me getting used to developing in the open and working through those initial oversights.

I learned a great amount working through the alpha pipeline quirks with you. Thanks for the great collaboration and for filing that upstream wgpu issue!
